### PR TITLE
add POST account functionality

### PR DESCRIPTION
**Pull request by turb0c0w**
*Issue:*
https://github.com/bcgov/entity/issues/21267
https://github.com/bcgov/entity/issues/21257
https://github.com/bcgov/entity/issues/21258

*Description of changes:*
  - new functionality for saving record on POST /account
  - new GET /registrations endpoint 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: strr-api/src/strr_api/services/registration_service.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+from strr_api import models, requests
+from strr_api.enums.enum import RegistrationStatus
+from strr_api.models import db
+
+
+class RegistrationService:
+    """Service to save and load regristration details from the database."""
+
+    @classmethod
+    def save_registration(cls, jwt_oidc_token_info, sbc_account_id, registration_request: requests.Registration):
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/services/registration_service.py:44]: payment will be added in a future sprint.

File: strr-api/src/strr_api/resources/registrations.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        description:
+    """
+
+    try:
+        registrations = RegistrationService.list_registrations(g.jwt_oidc_token_info)
+        return (
+            jsonify([Registration.from_db(registration).model_dump(mode="json") for registration in registrations]),
+            HTTPStatus.OK,
+        )
+    except AuthException as auth_exception:
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/resources/registrations.py:78]: yes

File: strr-api/src/strr_api/services/registration_service.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        user.email = registration_request.primaryContact.details.emailAddress
+        user.preferredname = registration_request.primaryContact.details.preferredName
+        user.phone_extension = registration_request.primaryContact.details.extension
+        user.fax_number = registration_request.primaryContact.details.faxNumber
+        user.phone_number = registration_request.primaryContact.details.phoneNumber
+        user.date_of_birth = registration_request.primaryContact.dateOfBirth
+
+        primary_contact = user
+        db.session.add(primary_contact)
+        db.session.commit()
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/services/registration_service.py:58]: good catch, this should have been .flush()

File: strr-api/src/strr_api/resources/account.py [hunk 4]
```diff
@@ -115,20 +118,92 @@ def create_account():
...
+        return jsonify(Registration.from_db(registration).model_dump(mode="json")), HTTPStatus.CREATED
+    except ValidationException as auth_exception:
+        return exception_response(auth_exception)
     except AuthException as auth_exception:
         return exception_response(auth_exception)
     except ExternalServiceException as service_exception:
         return exception_response(service_exception)
 
 
+def validate_and_format_canadian_postal_code(country: str, province: str, postal_code: str):
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/resources/account.py]: moved to RegistrationRequestValidator.py

File: strr-api/src/strr_api/resources/account.py [hunk 5]
```diff
@@ -115,20 +118,92 @@ def create_account():
...
         if not valid:
-            return error_response("Invalid request", HTTPStatus.BAD_REQUEST, errors)
+            raise ValidationException(message=errors)
 
-        name = json_input.get("name")
-        mailing_address = json_input.get("mailingAddress")
-        new_user_account = AuthService.create_user_account(token, name, mailing_address)
-        return jsonify(new_user_account), HTTPStatus.CREATED
+        registration_request = RegistrationRequest(**json_input)
+        selected_account = registration_request.selectedAccount
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/resources/account.py]: moved to RegistrationRequestValidator.py

File: strr-api/src/strr_api/responses/RegistrationResponse.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            status=source.status.name,
+            primaryContact=Contact(
+                name=ContactName(
+                    firstName=source.rental_property.property_manager.primary_contact_user.firstname,
+                    middleName=source.rental_property.property_manager.primary_contact_user.middlename,
+                    lastName=source.rental_property.property_manager.primary_contact_user.lastname,
+                ),
+                dateOfBirth=source.rental_property.property_manager.primary_contact_user.date_of_birth,
+                details=ContactDetails(
+                    preferredName=source.rental_property.property_manager.primary_contact_user.preferredname,
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/responses/RegistrationResponse.py:117]: Primary Contact is enforced as mandatory in the json schema. Updated the model to reflect that its required.

File: strr-api/src/strr_api/resources/account.py [hunk 7]
```diff
@@ -115,20 +118,92 @@ def create_account():
...
+        if registration_request.registration.secondaryContact:
+            registration_request.registration.secondaryContact.mailingAddress.postalCode = (
+                validate_and_format_canadian_postal_code(
+                    registration_request.registration.secondaryContact.mailingAddress.country,
+                    registration_request.registration.secondaryContact.mailingAddress.province,
+                    registration_request.registration.secondaryContact.mailingAddress.postalCode,
+                )
+            )
+
+        registration = RegistrationService.save_registration(
```
[the remark this answers is not in the document]
Comment (TylerZale) [reply strr-api/src/strr_api/resources/account.py]: Currently the endpoint is both (potentially) creating SBC user accounts and also saving registrations.

Based on your feedback I have moved it to POST /registrations and I believe in a future sprint we will likely split out the SBC account portion into its own endpoint.

File: strr-api/src/strr_api/services/registration_service.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        user.phone_number = registration_request.primaryContact.details.phoneNumber
+        user.date_of_birth = registration_request.primaryContact.dateOfBirth
+
+        primary_contact = user
+        db.session.add(primary_contact)
+        db.session.flush()
+        db.session.refresh(primary_contact)
+
+        if registration_request.secondaryContact:
+            secondary_contact = models.User(
```
[the remark this answers is not in the document]
Comment (turb0c0w) [reply strr-api/src/strr_api/services/registration_service.py:63]: There is no edit support currently, that work might come in a later sprint. Currently this method is the only time the data is saved, so there would be no orphaned records. That may change in future work.

**Review by trevoratindustrio — APPROVED** [review]

**Pull request merged**